### PR TITLE
i18n: add localized-file matching replacements check

### PR DESCRIPTION
**Pull request by brendankenny**
fixes https://github.com/GoogleChrome/lighthouse/pull/9846#issuecomment-542447233

> parse the english strings for markdown links / code snippets, and ensure each language has the same number

- `collect-strings.js`:
  - collects all UIStrings and brings them into `en-US.json` (and `en-XL.json`)
  - checks all other language files and deletes any strings with replacements that don't match the newly collected strings
https://github.com/GoogleChrome/lighthouse/blob/21c8d2f4f3d0bfc63e22d96bdaac73dad71019fd/lighthouse-core/scripts/i18n/prune-obsolete-lhl-messages.js#L108-L111

- `yarn i18n:checks` runs `collect-strings.js`, then
  - fails if the first bullet point above reveals a diff.

This PR asserts the second `collect-strings.js` bullet point by failing if there's a diff anywhere in the `locales/` directory (thereby failing if any language's strings had different replacements, different number of replacements, or different replacement arguments than the canonical versions).

### CODE REVIEW

File: lighthouse-core/scripts/i18n/assert-strings-collected.sh [hunk 1]
```diff
@@ -18,25 +20,13 @@ colorText() {
...
 set -x
 node "$lhroot_path/lighthouse-core/scripts/i18n/collect-strings.js" || exit 1
 set +x
 
-cp "$collectedstringsPath" "$freshstringsPath"
-
-colorText "Diff'ing golden strings against the fresh strings" "$purple"
-git --no-pager diff --color=always --no-index "$currentstringsPath" "$freshstringsPath"
+colorText "Diff'ing committed strings against the fresh strings" "$purple"
+git --no-pager diff --color=always --exit-code "$lhroot_path/lighthouse-core/lib/i18n/locales/"
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/assert-strings-collected.sh]: does the missing `--exit-code` imply that the `retVal` of the script before was always `0`?

**Comment by brendankenny** [comment]
> can this actually run on every commit though?

yeah. The [pruning step in `collect-strings`](https://github.com/GoogleChrome/lighthouse/blob/21c8d2f4f3d0bfc63e22d96bdaac73dad71019fd/lighthouse-core/scripts/i18n/collect-strings.js#L588) already deletes locale strings that don't match the current string, we just don't have a check anywhere that asserts you've run that before committing (at least as far as I can tell). The situation is already self correcting--next time someone runs `yarn update:sample-json` or other commands that collect strings, they'll get those deletions in their commit--so this is just asserting it on the commit where locale files are actually changed.

> wouldn't we start failing builds as soon as our locales get out of date and we're waiting on tc?

that's what the pruning step (#9598) added. It basically declared that if a locale string is out of date enough to have mismatched replacements it's not worth keeping around, so it's deleted and the english string is used until we get a new translation (the real impetus was that we'd get an error when we tried to do that replacement and it wasn't worth keeping some backward compatibility set of replacements around)

File: lighthouse-core/scripts/i18n/assert-strings-collected.sh (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/assert-strings-collected.sh]: > does the missing `--exit-code` imply that the `retVal` of the script before was always `0`?

I checked and it is working correctly on master for changes in `en-US.json`, so I think maybe the `--no-index` flag also opts into `--exit-code`? I can't find docs for that, though.

Without `--no-index` *and* without `--exit-code` this definitely always exits with `0`, so ¯\\_(ツ)_/¯

**Comment by brendankenny** [comment]
> fixes [#9846 (comment)](https://github.com/GoogleChrome/lighthouse/pull/9846#issuecomment-542447233)
> 
> > parse the english strings for markdown links / code snippets, and ensure each language has the same number

@exterkamp rightly points out this PR only catches half the story, since pruning takes place on LHL files, not CTC files, so some replacements have already been done.

However, this check plus the checks that already take place when baking do make this somewhat theoretical, but we could e.g. further expand [_ctcSanityChecks](https://github.com/GoogleChrome/lighthouse/blob/e3fcd6eba3b99b6e2314b20f3be8cd2261f9d59c/lighthouse-core/scripts/i18n/collect-strings.js#L343) and run it on input to baking so new language files get those checks on import for a new translations PR.

**Pull request merged**